src/Moq.Analyzers/ConstructorArgumentsShouldMatchAnalyzer.cs

```diff
@@ -1,252 +1,487 @@
+            argumentList = invocationExpressionSyntax.ArgumentList;
+        }
 
-            // REVIEW: Assert the MockingBehavior is the first argument in this case
-            // Skip first argument if it is not vararg - typically it is MockingBehavior argument
-            IEnumerable<ArgumentSyntax>? constructorArguments = objectCreation
-                                                                    .ArgumentList?
-                                                                    .Arguments
-                                                                    .Skip(varArgsConstructorParameterIndex == 0 ? 0 : 1);
+        VerifyMockAttempt(context, returnType, argumentList, hasMockBehavior);
+    }
 
-            if (mockedTypeSymbol.IsAbstract)
-            {
-                // Issue #1: Currently detection does not work well for abstract classes because they cannot be instantiated
+    /// <summary>
+    /// Analyzes when a Mock`1 object is created to verify the provided constructor arguments
+    /// match an existing constructor of the mocked class.
+    /// </summary>
+    /// <param name="context">The context.</param>
+    private void AnalyzeNewObject(SyntaxNodeAnalysisContext context)
+    {
+        ObjectCreationExpressionSyntax newExpression = (ObjectCreationExpressionSyntax)context.Node;
 
-                // The mocked symbol is abstract, so we need to check if the constructor arguments match the abstract class constructor
-                AnalyzeAbstract(context, constructorArguments, mockedTypeSymbol, objectCreation);
-            }
-            else
-            {
-                AnalyzeConcrete(context, constructorArguments, objectCreation, genericName);
-            }
+        GenericNameSyntax? genericNameSyntax = GetGenericNameSyntax(newExpression.Type);
+        if (genericNameSyntax == null)
+        {
+            return;
         }
-    }
 
-    private static void AnalyzeConcrete(
-        SyntaxNodeAnalysisContext context,
-        IEnumerable<ArgumentSyntax>? constructorArguments,
-        ObjectCreationExpressionSyntax objectCreation,
-        GenericNameSyntax genericName)
-    {
-        if (constructorArguments != null
-            && IsConstructorMismatch(context, objectCreation, genericName, constructorArguments))
+        // Quick check
+        if (!string.Equals(
+                genericNameSyntax.Identifier.ValueText,
+                WellKnownTypeNames.MockName,
+                StringComparison.Ordinal))
         {
-            Diagnostic diagnostic = objectCreation.ArgumentList.CreateDiagnostic(Rule);
-            context.ReportDiagnostic(diagnostic);
+            return;
         }
+
+        // Full check
+        SymbolInfo symbolInfo = context.SemanticModel.GetSymbolInfo(newExpression, context.CancellationToken);
+
+        if (symbolInfo.Symbol is not IMethodSymbol mockConstructorMethod
+            || mockConstructorMethod.MethodKind != MethodKind.Constructor
```

Comment:
nit: Can use your `IsConstructor` extension method here

Reply:
The test on L263 requires the extracted ctor symbol. I want to clean up a bunch of redundancies in the code like this. I'll take a pass on another PR